File: app/src/main/java/com/sysu/edu/view/PreferenceAdapter.java
package com.sysu.edu.view;

import static android.text.TextUtils.isEmpty;
import static com.sysu.edu.api.CommonUtil.toStringOrDefault;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.alibaba.fastjson2.JSONObject;
import com.sysu.edu.R;
import com.sysu.edu.databinding.ItemPreferenceBinding;

import java.util.List;
import java.util.stream.IntStream;

public class PreferenceAdapter extends RecyclerAdapter<JSONObject> {
    
    boolean hideNull = false;
    
    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new RecyclerView.ViewHolder(ItemPreferenceBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false).getRoot()) {
        };
    }
    
    public void set(List<Integer> titles, List<String> contents, List<Integer> icons, Context context) {
        clear();
        IntStream.range(0, titles.size()).forEach(i -> add(context.getString(titles.get(i)), contents.get(i), icons.get(i)));
    }
    
    public void add(String title, String content, Integer icon) {
        add(JSONObject.of("title", title, "content", content, "icon", icon));
        notifyItemInserted(getItemCount() - 1);
    }
    
    public void set(List<String> titles, List<String> contents, List<Integer> icons) {
        clear();
        IntStream.range(0, titles.size()).forEach(i -> add(titles.get(i), contents.get(i), icons.get(i)));
    }
    
    public void setHideNull(boolean hideNull) {
        this.hideNull = hideNull;
    }
    
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        int pos = holder.getBindingAdapterPosition();
        ItemPreferenceBinding binding = ItemPreferenceBinding.bind(holder.itemView);
        JSONObject item = get(position);
        binding.itemTitle.setText(toStringOrDefault(item.getString("title")));
        binding.itemContent.setText(isEmpty(item.getString("content")) ? holder.itemView.getContext().getString(R.string.none) : item.getString("content"));
        binding.getRoot().setOnClickListener(_ -> {
            // params.toast(titles.get(pos) + ": " + contents.get(pos));
        });
        binding.itemContent.setVisibility(hideNull && isEmpty(item.getString("content")) ? View.GONE : View.VISIBLE);
        if (item.getInteger("icon") != null)
            binding.itemIcon.setImageResource(item.getInteger("icon"));
//        else binding.itemIcon.setImageResource(R.drawable.account);
        binding.getRoot().updateAppearance(pos, getItemCount());
        super.onBindViewHolder(holder, position);
    }
}

File: app/src/main/java/com/miyuyan/sysuer/view/PreferenceAdapter.kt
package com.miyuyan.sysuer.view

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.alibaba.fastjson2.JSONObject
import com.miyuyan.sysuer.R
import com.miyuyan.sysuer.databinding.ItemPreferenceBinding

class PreferenceAdapter : RecyclerAdapter<JSONObject>() {
	var hideNull: Boolean = false
	override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
		return object : RecyclerView.ViewHolder(ItemPreferenceBinding.inflate(LayoutInflater.from(
			parent.context), parent, false).root) {}
	}
	
	fun set(titles: MutableList<Int?>,
	        contents: MutableList<String?>,
	        icons: MutableList<Int?>,
	        context: Context) {
		clear()
		titles.indices.forEach { add(context.getString(titles[it]!!), contents[it], icons[it]) }
	}
	
	fun add(title: String?, content: String?, icon: Int?) {
		add(JSONObject.of("title", title, "content", content, "icon", icon))
	}
	
	fun set(titles: MutableList<String?>,
	        contents: MutableList<String?>,
	        icons: MutableList<Int?>) {
		clear()
		titles.indices.forEach {
			add(titles.getOrNull(it), contents.getOrNull(it), icons.getOrNull(it))
		}
	}
	
	override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
		//val pos = holder.getBindingAdapterPosition()
		val item = get(position)
		val content = item.getString("content", "")
		ItemPreferenceBinding.bind(holder.itemView).apply {
			itemTitle.text = item.getString("title", "")
			itemContent.text = if (content.isNullOrEmpty()) holder.itemView.context.getString(R.string.none) else content
			root.setOnClickListener {}
			itemContent.visibility = if (hideNull && content.isNullOrEmpty()) View.GONE else View.VISIBLE
			if (item.getInteger("icon") != null) itemIcon.setImageResource(item.getInteger("icon")) //        else itemIcon.setImageResource(R.drawable.account);
			root.updateAppearance(position, itemCount)
		}
		super.onBindViewHolder(holder, position)
	}
}
